frostsnap_core/src/device.rs

```diff
@@ -1,36 +1,83 @@
 use core::num::NonZeroU32;
 
+use crate::symmetric_encryption::{Ciphertext, SymmetricKey};
+use crate::tweak::Xpub;
 use crate::{
-    message::*, ActionError, CheckedSignTask, Error, FrostKeyExt, MessageResult, SessionHash,
-    NONCE_BATCH_SIZE,
+    message::*, AccessStructureId, ActionError, CheckedSignTask, CoordShareDecryptionContrib,
+    Error, KeyId, MessageResult, SessionHash, NONCE_BATCH_SIZE,
 };
-use crate::{DeviceId, KeyId};
+use crate::{Appkey, DeviceId};
+use alloc::boxed::Box;
 use alloc::{
-    collections::{BTreeMap, BTreeSet},
+    collections::{BTreeMap, BTreeSet, VecDeque},
     string::String,
     vec::Vec,
 };
 use rand_chacha::ChaCha20Rng;
 use schnorr_fun::binonce;
 use schnorr_fun::frost::chilldkg::encpedpop;
-use schnorr_fun::frost::{PairedSecretShare, PartyIndex};
+use schnorr_fun::frost::{PairedSecretShare, PartyIndex, SecretShare};
 use schnorr_fun::fun::KeyPair;
 use schnorr_fun::{
     binonce::{Nonce, NonceKeyPair},
     frost::{self},
     fun::{derive_nonce_rng, prelude::*, Tag},
     nonce, Message,
 };
-use sha2::digest::{FixedOutput, Update};
 
 use sha2::Sha256;
 
 #[derive(Clone, Debug)]
 pub struct FrostSigner {
     keypair: KeyPair,
-    secret_shares: BTreeMap<KeyId, PairedSecretShare>,
+    keys: BTreeMap<KeyId, KeyData>,
     action_state: Option<SignerState>,
     nonce_counter: u64,
+    mutations: VecDeque<Mutation>,
+}
+
+#[derive(Clone, Debug)]
+pub struct KeyData {
+    access_structures: BTreeMap<AccessStructureId, AccessStrucureData>,
+    #[allow(dead_code)] // We'll use this soon
+    purposes: BTreeSet<KeyPurpose>,
```

Comment:
The plan for this is so the device can tell the coordinator what the purpose of the share is. Presumably a nostr app should ignore bitcoin shares.